### ISSUE TITLE
Issues with inheritance using babel and ES6 classes

### ISSUE THREAD

**Issue by stalniy** [js-data/js-data#327]
**Description**:
I'm trying to use custom `recordClass` inside mapper. For doing this I extend `JSData.Record` in order to have all the niceties of schema validation. But validation doesn't work.

**Example**: https://jsfiddle.net/wyumc5ao/2/

**Where is the issue?**:
The issue is in `__super__` accessor which is defined by `extend` function. This property is inherited by child class (i.e., `Task` in the example) from `Record` class. Babel doesn't set `__super__` property on the constructor and then `utils.getSuper(...) === Record` fails because `utils.getSuper` returns `Component` and not `Record` as it should.

**Possible fix**:
Add additional check for `hasOwnProperty('__super__'). Something like this:

```js
    getSuper: function getSuper(instance, isCtor) {
      var ctor = isCtor ? instance: instance.constructor;
 
      if (!ctor.hasOwnProperty('__super__')) {
         Object.defineProperty(ctor, '__super__', { configurable: true, value: Object.getPrototypeOf(ctor) || ctor.__proto__ }); // eslint-disable-line
      }

      return ctor.__super__ // eslint-disable-line; 
    },
```

**Conclusion**:
If you are ok with the fix I'll send pull-request on Sunday

**Comment by jmdobry** [comment]
It's okay that `__super__` isn't set with ES2015 class inheritance. `getSuper` will look for `__super__` (which is present if you use `.extend`) or `Object.getPrototypeOf(instance.constructor)` (which is present if you use `class` syntax).

Validation doesn't work because according to JSON-schema, you need `type: 'object'` as a top-level property of the schema. It would probably be best if JSData adds that for you.

So in your JSFiddle, if you add `type: 'object'` to your schema, validation will work. e.g.

```js
schema: {
  type: 'object',
  properties: {...}
}
```

**Comment by stalniy** [comment]
`getSuper` returns wrong result for es2015 inheritance because parent class is set as prototype for child class.

That means all static properties are inherited from parent class which has `__super__` property. And when you do `Child.__super__` it will fallback to `Parent.__super__` which eventually returns a truthy result. So, Object.getPrototyOf is never called

**Comment by jmdobry** [comment]
> That means all static properties are inherited from parent class which has `__super__` property

That makes sense.

**Comment by jmdobry** [comment]
Fixed in 3.0.0-beta.5

**Comment by stalniy** [comment]
Thanks a lot!
